### ISSUE TITLE
[feat] OIDC

### ISSUE THREAD

**Issue by ntimo** [gregtwallace/certwarden#9]
Hello,
What do you think about adding oidc single sign on? This would be super cool, since then it’s possible to have sso even for issuing let’s encrypt certs.

best wishes,
And thanks for creating this awesome projekt.

**Comment by gregtwallace** [comment]
Do you have any background in this area or links to guides/documentation?  I think SSO could be a good feature add but it is outside my wheelhouse.

**Comment by ntimo** [comment]
I never actually implemented this my self, so no. But just googled and found this library that looks pretty good https://github.com/zitadel/oidc

**Comment by gregtwallace** [comment]
Are you still interested in this? What sso product are you trying to integrate with?

I’m redoing my home lab currently and might look into this soonish.

**Comment by ntimo** [comment]
Yes defiantly. I am using using https://www.authelia.com/ for OIDC / SSO.

**Comment by sjafferali** [comment]
+1 also using authelia + a reverse proxy, so also wanting a solution for this.

**Comment by maverick2478** [comment]
> +1 also using authelia + a reverse proxy, so also wanting a solution for this.

same here. looking forward to a solution with authelia + traefik for example. anyway thumbs up to the Greg for this remarkarble application. :)

**Comment by Tivin-i** [comment]
Maybe this will assist? There is an example for a go application.

https://github.com/PacktPublishing/Keycloak-Identity-and-Access-Management-for-Modern-Applications/tree/master/ch7

https://github.com/coreos/go-oidc

This is from the book "Keycloak - Identity and Access Management for Modern Applications: Harness the power of Keycloak, OpenID Connect, and OAuth 2.0 to secure applications, Second Edition"

**Comment by tompaah** [comment]
Right now about 10,000 research and education institutions in the European [GÉANT](https://about.geant.org/about/) partnership network are all looking for a new cert management solution, since GÉANTs certificate partner Sectigo unexpectedly [announced the termination of the agreement](https://www.ca.kit.edu/english/index_285.php) to provide certificates to the 10,000 members in the network. Most are discussing moving to Let's Encrypt but this required automation and management (since the renewal time drops from 365 days to 90) which most also happen to lack.

I'd bet Cert Warden would be very attractive if it had OpenID based login capabilities, almost all R&E institutions already have these SSO methods in place. As a second alternative, either LDAP or proxy-based login might do.

**Comment by SirNerdBear** [comment]
An alternative to SSO support would be a configuration option that supports turning on Basic AUTH, as opposed to a form. Then a reverse proxy can be configured to pass Authorization header and use an authorization forward for SSO.

**Comment by gregtwallace** [comment]
I've made pretty good progress architechting and coding this locally. I anticipate a test Docker build (which will only support OIDC) in the next week or two, which will need testing and feedback.

Once all the kinks are worked out I'll figure out how to merge into the main branch.

**Comment by gregtwallace** [comment]
Happy new year!

Here is a docker build with OIDC added.

**Warning: There are other changes in this build which will upgrade your database version. Backup first!**

If you have any issues or errors, they will be in the log files. The frontend GUI will probably only say `unauthorized` for almost all issues since this process is pre-authentication. Turn on debug logging and consult your log files if you're having issues. 

Also note, the `logout` function does not log you out of the Idp.

Please report back successes and issues.


You will need to modify your CertWarden config manually:
`
auth:
    oidc:
        issuer_url: 'https://yourIdp.example.com/'
        client_id: 'some-id'
        client_secret: 'some-secret'
        api_redirect_uri: 'https://your-certwarden-api.example.com:4055/certwarden/api/v1/app/auth/oidc/callback'
`
Notes:
`issuer_url` is extremely picky and requires the leading scheme `https://` as well as the trailing slash.
`api_redirect_url` should be this exact url, just replace the host portion and port with your Cert Warden API host.

Your Idp should be configured to:
Support audience `your-certwarden-api.example.com:4055/certwarden/api`
Support scope `certwarden:superuser` and add this permission to users who should have access (ensure your Idp is enforcing RBAC!)
Add the callback URL specified in the CW config above.
Add / support the `offline_access` scope.


Sorry for the weird split files. GitHub has a file size limit. You will have to unzip each file and then use WinRAR to extract the image. Alternatively, build it yourself: `docker build . --build-arg=BACKEND_VERSION=v0.24.0-a2 --build-arg=FRONTEND_VERSION=v0.24.0-a2 -t certwarden:v0.24.0-a2`

[cw0.24.0-a2.part1.zip](https://github.com/user-attachments/files/18286191/cw0.24.0-a2.part1.zip)
[cw0.24.0-a2.part2.zip](https://github.com/user-attachments/files/18286194/cw0.24.0-a2.part2.zip)

**Comment by tompaah** [comment]
I've successfully tested the image with two IdPs capable of OIDC;

1. **Active Directory Federation Services** on Windows Server 2022 (ADFS "version 6.0"). Works well like any other OpenID capable application. Just had to add both `offline_access` and `certwarden:superuser` as scopes. I'm using an ADFS access policy with MFA enabled and it triggered as it should.
2. **Authentik** 2024.12.1. Also without problems, had to add the `offline_access` scope to the provider.

If you'd be happy for it I could document what configuration steps are needed for these two IdPs so that can go in the official documentation whenever the feature is ready for release.

Since the application now effectively moves from single-user mode to multi-user, it would be very handy if it could have a field in the database that notes what user created the object (key, certificate, acme orders, - servers and providers). In larger environments you get stuck with weird objects after a while that nobody knows who done it and why it is there, it's very helpful at least to know who did it. 

Rights management could also be implemented but this'd be a bigger task and since the application is mostly a superuser application anyway, it's sufficient to have IdP level access control; either you can use the application or you cannot.

**Comment by gregtwallace** [comment]
Released in v0.24.0.
